api/handlers/tag_handler_test.go

```diff
@@ -1,98 +1,363 @@
+}
+
+func TestFindTags_byFuzz(t *testing.T) {
+	store := getTestTagStore(t, testTags)
+	handler := NewTagHandler(store)
+
+	cases := []HandlerTestCase{}
+	for _, f := range []string{"d", "e", "j", "l", "s", "t"} {
+		fuzz := f
+
+		cases = append(cases, HandlerTestCase{
+			Name: fuzz,
 			Request: &TestRequest{
-				Query: "key1=val1&key2=val2",
+				Query: fmt.Sprintf("fuzz=%s", fuzz),
+			},
+			Run: func(r *testutils.Reporter, _ interface{}, req *restful.Request, resp *restful.Response, read Readf) {
+				handler.FindTags(req, resp)
+
+				var tags []models.EntityWithTags
+				read(&tags)
+
+				r.AssertEqual(len(tags), 2)
 			},
-			Setup: func(ctrl *gomock.Controller) interface{} {
-				tagDataMock := mock_data.NewMockTagData(ctrl)
+		})
+	}
+
+	RunHandlerTestCases(t, cases)
+}
 
-				params := map[string]string{
-					"key1": "val1",
-					"key2": "val2",
+func TestFindTags_byID(t *testing.T) {
+	store := getTestTagStore(t, testTags)
+	handler := NewTagHandler(store)
+
+	cases := []HandlerTestCase{}
+	for _, i := range []string{"d1", "d2", "e1", "e2", "j1", "j2", "l1", "l2", "s1", "s2", "t1", "t2"} {
+		id := i
+
+		cases = append(cases, HandlerTestCase{
+			Name: id,
+			Request: &TestRequest{
+				Query: fmt.Sprintf("id=%s", id),
+			},
+			Run: func(r *testutils.Reporter, _ interface{}, req *restful.Request, resp *restful.Response, read Readf) {
+				handler.FindTags(req, resp)
+
+				var tags []models.EntityWithTags
+				read(&tags)
+
+				r.AssertEqual(len(tags), 1)
+				r.AssertEqual(tags[0].EntityID, id)
+			},
+		})
+	}
+
+	RunHandlerTestCases(t, cases)
+}
+
+func TestFindTags_byType(t *testing.T) {
+	store := getTestTagStore(t, testTags)
+	handler := NewTagHandler(store)
+
+	cases := []HandlerTestCase{}
+	for _, t := range []string{"deploy", "environment", "job", "load_balancer", "service", "task"} {
+		typeName := t
+
+		cases = append(cases, HandlerTestCase{
+			Name: typeName,
+			Request: &TestRequest{
+				Query: fmt.Sprintf("type=%s", typeName),
+			},
+			Run: func(r *testutils.Reporter, _ interface{}, req *restful.Request, resp *restful.Response, read Readf) {
+				handler.FindTags(req, resp)
+
+				var tags []models.EntityWithTags
+				read(&tags)
+
+				r.AssertEqual(len(tags), 2)
+				for _, tag := range tags {
+					r.AssertEqual(tag.EntityType, typeName)
 				}
+			},
+		})
+	}
+
+	RunHandlerTestCases(t, cases)
+}
+
+func TestFindTags_byVersion(t *testing.T) {
+	store := getTestTagStore(t, []*models.Tag{
+		{EntityID: "d1", EntityType: "deploy", Key: "version", Value: "1"},
+		{EntityID: "d2", EntityType: "deploy", Key: "version", Value: "2"},
+		{EntityID: "d3", EntityType: "deploy", Key: "version", Value: "3"},
+	})
+
+	handler := NewTagHandler(store)
+
+	cases := []HandlerTestCase{
+		{
+			Name: "2",
+			Request: &TestRequest{
+				Query: "version=2",
+			},
+			Run: func(r *testutils.Reporter, _ interface{}, req *restful.Request, resp *restful.Response, read Readf) {
+				handler.FindTags(req, resp)
```

Comment:
[nit] The "call method and marshal" workflow could be made into a single helper function since all handler methods are `func(req, resp)`:
```
Run: func(r *testutils.Reporter, target interface{}, calllAndMarshal Readf) {
    var tags []models.EntityWithTags
    callAndMarshal(handler.FindTags, &tags)

    r.AssertEqual(len(tags), 2)
    ...
}
```